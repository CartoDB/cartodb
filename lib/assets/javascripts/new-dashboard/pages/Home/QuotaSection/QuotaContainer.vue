<template>
  <div class="quota-container">
    <div class="quota-headers">
      <div class="quota-cell cell--main">
        <h3 class="title is-caption quota-title">{{title}}</h3>
      </div>

      <div class="quota-labels">
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey">{{ $t('QuotaSection.header.remaining') }}</span>
        </div>
        <div class="quota-cell cell--medium cell--mobile">
          <span class="text is-small is-txtSoftGrey">{{ $t('QuotaSection.header.used') }}</span>
        </div>
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey quota-total" :class="{'is-active' : perMonth }">{{$t(`QuotaSection.header.total`)}}</span>
          <span class="text is-small is-txtSoftGrey quota-month" :class="{'is-active' : perMonth }">&nbsp;{{$t(`QuotaSection.header.perMonth`)}}</span>
        </div>
        <div class="quota-cell cell--small"></div>
      </div>
    </div>
    <div class="quota-info">
      <slot></slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'QuotaContainer',
  props: {
    title: String,
    perMonth: Boolean
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quota-headers {
  display: flex;
  justify-content: space-between;
}

.quota-title {
  margin-bottom: 8px;
}

.quota-labels {
  display: flex;
}

.quota-info {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: $white;
}

.quota-cell {
  display: flex;
  align-items: center;
  padding: 0 10px;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }
}

.cell--mobile {
  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.cell--main {
  min-width: 300px;

  @media (max-width: $layout-mobile) {
    min-width: unset;
  }
}

.cell--medium {
  width: 120px;
}

.cell--small {
  width: 58px;

  @media (max-width: $layout-tablet) {
    display: none;
  }
}

.quota-month {
  display: none;
}

.is-active {
  &.quota-total {
    @media (max-width: $layout-tablet) and (min-width: $layout-mobile) {
      &::after {
        content: '*';
      }
    }
  }

  &.quota-month {
    display: flex;

    @media (max-width: $layout-tablet) and (min-width: $layout-mobile) {
      display: none;
    }

    &::after {
      content: '*';
    }
  }
}

</style>
