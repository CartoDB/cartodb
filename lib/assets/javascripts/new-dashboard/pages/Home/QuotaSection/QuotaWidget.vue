<template>
  <div class="quota-widget" @mouseover="onMouseOver" @mouseleave="onMouseLeave">
    <div class="quota-main">
      <div class="quota-cell cell--title">
        <h4 class="text is-caption is-semibold is-txtGrey">{{name}}</h4>
      </div>
      <div class="quota-cell cell--large">
        <div class="progressbar">
            <div :class="`progressbar progressbar--${getStatusBar}`"  :style="{width: `${getUsedPercent}%`}">
            </div>
        </div>
      </div>
    </div>

    <div class="quota-data">
      <div class="quota-cell cell--medium">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(remainingQuota)}} {{unit}}</span>
      </div>
      <div class="quota-cell cell--medium cell--mobile">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(usedQuota)}} {{unit}}</span>
      </div>
      <div class="quota-cell cell--medium">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(availableQuota)}} {{unit}}</span>
      </div>
      <div class="quota-help cell--small">
        <a href=""><img svg-inline class="quota-image" :class="{'is-active': active}" src="../../../assets/icons/common/question-mark.svg"/></a>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'QuotaCard',
  props: {
    name: String,
    availableQuota: Number,
    usedQuota: Number,
    unit: String,
    billingPeriod: String
  },
  data: function () {
    return {
      active: false
    };
  },
  computed: {
    remainingQuota () {
      return this.availableQuota - this.usedQuota;
    },
    getUsedPercent () {
      if (this.availableQuota === 0) {
        return 100;
      }
      const width = (this.usedQuota / this.availableQuota) * 100;
      return width;
    },
    getStatusBar () {
      const usedPercent = this.getUsedPercent;
      if (usedPercent > 90) {
        return 'problem';
      } else if (usedPercent > 60) {
        return 'warning';
      } else {
        return 'good';
      }
    }
  },
  methods: {
    roundOneDecimal (number) {
      return Math.round(number * 10) / 10;
    },
    onMouseOver () {
      this.active = true;
    },
    onMouseLeave () {
      this.active = false;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quota-widget {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 80px;
  cursor: pointer;

  &:not(:last-of-type) {
    border-bottom: 1px solid $softblue;
  }
}

.quota-main {
  display: flex;
  width: 100%;
}

.quota-data {
  display: flex;
}

.quota-cell {
  display: flex;
  align-items: center;
  padding: 0 10px;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }
}

.cell--mobile {
  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.quota-help {
  display: flex;
  justify-content: flex-end;
  padding-right: 16px;
}

.quota-image {
  display: none;

  &.is-active {
    display: flex;
  }
}

.cell--title {
  width: 110px;
  margin-left: 36px;
}

.cell--large {
  flex-grow: 1;
  flex-shrink: 1;
  min-width: 120px;
  margin-right: 40px;

  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.cell--medium {
  width: 120px;
}

.cell--small {
  width: 58px;

  @media (max-width: $layout-tablet) {
    display: none;
  }
}

.progressbar {
  width: 100%;
  max-width: 240px;
  height: 8px;
  border-radius: 4px;
  background-color: $light-grey;

  &.progressbar--good {
    background-color: $good-state;
  }

  &.progressbar--warning {
    background-color: $warning-state;
  }

  .progressbar--problem {
    background-color: $error-state;
  }
}

</style>
