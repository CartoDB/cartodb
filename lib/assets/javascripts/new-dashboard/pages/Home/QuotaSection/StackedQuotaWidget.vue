<template>
  <div class="quota-widget-wrapper">
    <div class="quota-widget" :class="{'is-disabled': isDisabled }" @mouseover="onMouseOver" @mouseleave="onMouseLeave">
      <div class="quota-main">
        <div class="quota-cell cell--title">
          <h4 class="text is-caption is-semibold is-txtGrey">{{name}}</h4>
        </div>
        <div class="quota-cell cell--large">
          <NotificationBadge  v-if="isDisabled">
            <div class="text is-small" v-html="$t('QuotaSection.upgrade', { path: upgradeUrl })"></div>
          </NotificationBadge>
          <div v-else class="progressbar">
            <Tooltip :text="tooltipText" position="bottom-right" hide-delay="0s" show-delay="1s">
              <template v-if="Array.isArray(usedQuota)">
                <div v-for="section in usedQuota"
                  :key="section.label"
                  :class="`progressbar ${qualitative ? `progressbar--${getStatusBar}` : 'progressbar--filled'}`"
                  :style="{backgroundColor: section.color, width: `${getUsedPercent(section.value)}%`}">
                </div>
              </template>
              <template v-else>
                <div
                  :class="`progressbar ${qualitative ? `progressbar--${getStatusBar}` : 'progressbar--filled'}`"
                  :style="{width: `${getUsedPercent(usedQuota)}%`}">
                </div>
              </template>
            </Tooltip>
          </div>
        </div>
      </div>

      <div class="quota-data">
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey" v-if="!unlimited">
            {{ formatToLocale ? getNumberInLocaleFormat(roundOneDecimal(remainingQuota)) : roundOneDecimal(remainingQuota) }} {{unit}}
          </span>
        </div>
        <div class="quota-cell cell--medium cell--mobile">
          <span class="text is-small is-txtSoftGrey" v-if="!unlimited">
            {{ formatToLocale ? getNumberInLocaleFormat(roundOneDecimal(totalUsedQuota)) : roundOneDecimal(totalUsedQuota) }} {{unit}}
          </span>
        </div>
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey" v-if="!unlimited">
            {{ formatToLocale ? getNumberInLocaleFormat(roundOneDecimal(totalQuota)) : roundOneDecimal(totalQuota) }} {{unit}}
          </span>
          <span class="text is-small is-txtSoftGrey" v-else>
            {{ formatToLocale ? getNumberInLocaleFormat(roundOneDecimal(totalUsedQuota)) : roundOneDecimal(totalUsedQuota) }} {{unit}}
          </span>
        </div>
        <div class="quota-help cell--small">
          <Tooltip :text="helpText" position="bottom-left" hide-delay="0s" show-delay="1s">
            <a :href="helpLink" v-if="helpLink" target= "_blank"><img svg-inline class="quota-image" :class="{'is-active': active}" src="../../../assets/icons/common/question-mark.svg"/></a>
          </Tooltip>
        </div>
      </div>
    </div>
    <div class="quota-legend" v-if="Array.isArray(usedQuota)">
      <div v-for="item in usedQuota"
        :key="item.label"
        class="legend-item text is-small is-semibold">
          <span class="legend-icon" :style="{backgroundColor: item.color}"></span>
          {{item.label}}
      </div>
    </div>
    <div class="quota-warning" v-if="showWarning">
      <p class="is-small is-bold">
        <img src="../../../assets/icons/apps/warning.svg" />
        {{$t('QuotaSection.quotaExceeded')}}
      </p>
    </div>
  </div>
</template>

<script>
import NotificationBadge from 'new-dashboard/components/NotificationBadge';
import Tooltip from 'new-dashboard/components/Tooltip/Tooltip';
import { mapState } from 'vuex';

const THRESHOLDS = {
  WARNING: 60,
  PROBLEM: 90
};

export default {
  name: 'StackedQuotaCard',
  props: {
    name: String,
    availableQuota: Number,
    unit: String,
    usedQuota: [Array, Number],
    billingPeriod: String,
    formatToLocale: {
      type: Boolean,
      default: true
    },
    helpText: String,
    helpLink: String,
    isDisabled: {
      type: Boolean,
      default: false
    },
    unlimited: {
      type: Boolean,
      default: false
    },
    qualitative: {
      type: Boolean,
      default: false
    }
  },
  components: {
    NotificationBadge,
    Tooltip
  },
  data: function () {
    return {
      active: false
    };
  },
  computed: {
    ...mapState({
      upgradeUrl: state => state.config.upgrade_url
    }),
    showWarning () {
      return this.remainingQuota <= 0;
    },
    tooltipText () {
      let tooltip = Array.isArray(this.usedQuota) ? `${this.usedQuota.map(quotaEntry => {
        return `${quotaEntry.label} ${this.getNumberInLocaleFormat(this.roundOneDecimal(quotaEntry.value))} ${this.unit}`;
      }).join('<br>')}` : null;

      return tooltip;
    },
    totalUsedQuota () {
      return !Array.isArray(this.usedQuota) ? this.usedQuota : this.usedQuota.reduce((acum, current) => {
        return acum + current.value;
      }, 0);
    },
    totalQuota () {
      return this.availableQuota;
    },
    remainingQuota () {
      return this.totalQuota - this.totalUsedQuota;
    },
    getStatusBar () {
      const usedPercent = this.getUsedPercent(this.totalUsedQuota);
      if (usedPercent > THRESHOLDS.PROBLEM) {
        return 'problem';
      } else if (usedPercent > THRESHOLDS.WARNING) {
        return 'warning';
      } else {
        return 'good';
      }
    }
  },
  methods: {
    getUsedPercent (used) {
      if (this.totalQuota === 0) {
        return 100;
      }
      const width = (used / this.totalQuota) * 100;

      return width;
    },
    roundOneDecimal (number) {
      return Math.round(number * 10) / 10;
    },
    getNumberInLocaleFormat (number) {
      return number.toLocaleString();
    },
    onMouseOver () {
      this.active = true;
    },
    onMouseLeave () {
      this.active = false;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.quota-widget-wrapper {
  width: 100%;
  padding: 28px 0;
  min-height: 80px;

  &:not(:last-of-type) {
    border-bottom: 1px solid $border-color;
  }

  .quota-warning {
    display: flex;
    padding-top: 20px;
    margin-left: 36px;

    p {
      display: flex;
      align-items: center;
      border-radius: 18px;
      padding: 8px 16px 8px 10px;
      background-color: $warning__bg-color-light;

      img {
        margin-right: 8px;
      }
    }

  }

  .quota-legend {
    display: flex;
    padding-top: 20px;
    margin-left: 206px;

    .legend-item {
      display: flex;
      align-items: center;

      &:not(:last-child) {
        margin-right: 14px;
      }

      .legend-icon {
        display: block;
        height: 8px;
        width: 8px;
        margin-right: 6px;
        border-radius: 50%;
      }
    }

  }
}

.quota-widget {
  display: flex;
  align-items: center;
  justify-content: space-between;

  &.is-disabled {
    background-color: rgba($color-primary--soft, 0.3);

    .cell--title,
    .quota-data {
      opacity: 0.3;
    }
  }
}

.quota-main {
  display: flex;
  width: 100%;
}

.quota-data {
  display: flex;
}

.quota-cell {
  display: flex;
  align-items: center;
  padding: 0 10px;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }
}

.cell--mobile {
  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.quota-help {
  display: flex;
  justify-content: flex-end;
  padding-right: 24px;
  cursor: pointer;
}

.quota-image {
  display: none;

  &.is-active {
    display: flex;
  }
}

.cell--title {
  width: 160px;
  margin-left: 36px;
}

.cell--large {
  flex-grow: 1;
  flex-shrink: 1;
  min-width: 120px;
  margin-right: 40px;

  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.cell--medium {
  width: 110px;
}

.cell--small {
  width: 58px;

  @media (max-width: $layout-tablet) {
    display: none;
  }
}

.progressbar {
  display: flex;
  overflow: hidden;
  flex: 0 0 auto;
  width: 100%;
  max-width: 260px;
  height: 8px;
  border-radius: 4px;
  background-color: $progressbar__bg-color;

  .tooltip-container {
    width: 100%;
  }

  &:not(:last-child) {
    border-radius: 4px 0 0 4px;
    border-right: 1px solid $progressbar__bg-color;
  }
  &:not(:first-child) {
    // border-radius: 0 4px 4px 0;
    border-radius: 0;
  }

  &.progressbar--filled {
    background-color: $color-primary--dark;
  }

  &.progressbar--good {
    background-color: $success__bg-color;
  }

  &.progressbar--warning {
    background-color: $warning__bg-color;
  }

  .progressbar--problem {
    background-color: $danger__bg-color;
  }
}

.warning-icon {
  margin-right: 5px;
}

</style>
