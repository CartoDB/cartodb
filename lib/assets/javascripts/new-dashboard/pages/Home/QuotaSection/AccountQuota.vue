<template>
  <QuotaContainer :title="$t(`QuotaSection.account`)" :perMonth="false">
    <QuotaWidget
      :name="$t(`QuotaSection.storage`)"
      :usedQuota="getAmountInUnit(usedStorage, amountExponent)"
      :availableQuota="getAmountInUnit(availableStorage, amountExponent)"
      :unit="getUnit(amountExponent)"
      :formatToLocale="false"
      :helpLink="storageHelpLink"/>

    <QuotaWidget
      v-if="hasTableLimits"
      :name="$t(`QuotaSection.publicMaps`)"
      :usedQuota="usedPublicMaps"
      :availableQuota="availablePublicMaps"/>

    <QuotaWidget
      v-if="hasPrivateMapsLimits"
      :name="$t(`QuotaSection.privateMaps`)"
      :usedQuota="usedPrivateMaps"
      :availableQuota="availablePrivateMaps"/>

    <QuotaWidget
      v-if="hasPublicMapLimits"
      :name="$t(`QuotaSection.datasets`)"
      :usedQuota="usedDatasets"
      :availableQuota="availableDatasets"/>

    <QuotaWidget
      v-if="hasApiKeysLimits"
      :name="$t(`QuotaSection.apiKeys`)"
      :usedQuota="usedApiKeys"
      :availableQuota="availableApiKeys"
      :isDisabled="hasApiKeysLimitsToZero"/>
  </QuotaContainer>
</template>

<script>
import { mapState } from 'vuex';
import QuotaWidget from './QuotaWidget';
import QuotaContainer from './QuotaContainer';
import { apiKeysTypes } from 'new-dashboard/core/constants/api-keys';
import * as Accounts from 'new-dashboard/core/constants/accounts';

export default {
  name: 'AccountQuota',
  components: {
    QuotaWidget,
    QuotaContainer
  },
  beforeMount () {
    this.getApiKeysTotal();
  },
  data () {
    return {
      usedApiKeys: 0
    };
  },
  computed: {
    ...mapState({
      availableStorage: state => state.user.quota_in_bytes,
      remainingStorage: state => state.user.remaining_byte_quota,
      availablePublicMaps: state => state.user.public_map_quota,
      linkMapsTotal: state => state.user.link_privacy_map_count,
      passwordMapsTotal: state => state.user.password_privacy_map_count,
      publicMapsTotal: state => state.user.public_privacy_map_count,
      availablePrivateMaps: state => state.user.private_map_quota,
      privateMapsTotal: state => state.user.private_privacy_map_count,
      mapsTotal: state => state.user.owned_visualization_count,
      availableDatasets: state => state.user.table_quota,
      usedDatasets: state => state.user.table_count,
      availableApiKeys: state => state.user.regular_api_key_quota,
      planAccountType: state => state.user.account_type
    }),
    usedStorage () {
      return this.availableStorage - this.remainingStorage;
    },
    amountExponent () {
      return this.getExpBaseTwo(this.availableStorage) - 10;
    },
    storageHelpLink () {
      return 'https://carto.com/help/your-account/your-disk-storage/';
    },
    hasTableLimits () {
      return Accounts.accountsWithTableLimits.includes(this.planAccountType);
    },
    hasPublicMapLimits () {
      return Accounts.accountsWithPublicMapLimits.includes(this.planAccountType);
    },
    hasPrivateMapsLimits () {
      return Accounts.accountsWithPrivateMapsLimits.includes(this.planAccountType);
    },
    hasApiKeysLimits () {
      return Accounts.accountsWithApiKeysLimits.includes(this.planAccountType);
    },
    hasApiKeysLimitsToZero () {
      return Accounts.accountsWithApiKeysLimitsToZero.includes(this.planAccountType);
    },
    usedPublicMaps () {
      return this.linkMapsTotal + this.passwordMapsTotal + this.publicMapsTotal;
    },
    usedPrivateMaps () {
      return this.mapsTotal - this.usedPublicMaps;
    },
    privateMapsTotal () {
      return this.mapsTotal - this.linkMapsTotal - this.passwordMapsTotal - this.publicMapsTotal;
    }
  },
  methods: {
    getExpBaseTwo (sizeInBytes) {
      if (sizeInBytes === 0) {
        return 0;
      }

      let exponent = 0;
      if (Math.log2) {
        exponent = Math.log2(sizeInBytes);
      } else {
        exponent = Math.log(sizeInBytes) * Math.LOG2E;
      }

      return Math.round(exponent / 10) * 10;
    },
    getUnit (exponent) {
      if (exponent < 10) {
        return 'B';
      } else if (exponent < 20) {
        return 'KB';
      } else if (exponent < 30) {
        return 'MB';
      } else if (exponent < 40) {
        return 'GB';
      } else if (exponent < 50) {
        return 'TB';
      } else if (exponent < 60) {
        return 'PB';
      } else if (exponent < 70) {
        return 'EB';
      } else {
        return '?';
      }
    },
    getAmountInUnit (number, exponent) {
      return number / Math.pow(2, exponent);
    },
    getApiKeysTotal () {
      this.$store.state.client.getApiKeys(apiKeysTypes.REGULAR,

        (err, _, data) => {
          if (err) {
            return;
          }
          this.usedApiKeys = data.count;
        }
      );
    }
  }
};
</script>
