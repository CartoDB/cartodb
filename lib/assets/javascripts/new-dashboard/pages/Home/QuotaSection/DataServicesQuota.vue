<template>
  <QuotaContainer :title="$t(`QuotaSection.dataServices`)" :perMonth="true">
    <QuotaWidget :name="$t(`QuotaSection.geocoding`)" :usedQuota="geocodingUsed" :availableQuota="geocodingAvailable" :helpLink="geocodingHelpLink"></QuotaWidget>
    <QuotaWidget :name="$t(`QuotaSection.isolines`)" :usedQuota="isolinesUsed" :availableQuota="isolinesAvailable" :helpLink="isolinesHelpLink"></QuotaWidget>
    <QuotaWidget :name="$t(`QuotaSection.routing`)" :usedQuota="routingUsed" :availableQuota="routingAvailable" :helpLink="routingHelpLink"></QuotaWidget>
  </QuotaContainer>
</template>

<script>
import { mapState } from 'vuex';
import QuotaWidget from './QuotaWidget';
import QuotaContainer from './QuotaContainer';

export default {
  name: 'DataServicesQuota',
  components: {
    QuotaWidget,
    QuotaContainer
  },
  computed: {
    ...mapState({
      geocodingUsed: state => state.user.geocoding.monthly_use,
      geocodingAvailable: state => state.user.geocoding.quota,
      routingUsed: state => state.user.mapzen_routing.monthly_use,
      routingAvailable: state => state.user.mapzen_routing.quota ? state.user.mapzen_routing.quota : 0,
      isolinesUsed: state => state.user.here_isolines.monthly_use,
      isolinesAvailable: state => state.user.here_isolines.quota
    }),
    geocodingHelpLink () {
      return 'https://carto.com/help/working-with-data/geocoding/';
    },
    isolinesHelpLink () {
      return 'https://carto.com/help/working-with-data/isolines/';
    },
    routingHelpLink () {
      return 'https://carto.com/help/working-with-data/routing/';
    }
  }
};
</script>
