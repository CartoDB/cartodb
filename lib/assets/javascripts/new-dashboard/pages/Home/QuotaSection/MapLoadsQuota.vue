<template>
  <QuotaContainer :title="$t(`QuotaSection.mapLoads`)" :perMonth="true">
    <StackedQuotaWidget
      :name="$t(`QuotaSection.quota`)"
      :usedQuota="mapViews"
      :availableQuota="mapViewsQuota"
      :formatToLocale="true"
      :helpLink="helpLink" qualitative/>
  </QuotaContainer>
</template>

<script>
import { mapState } from 'vuex';
import QuotaContainer from './QuotaContainer';
import StackedQuotaWidget from './StackedQuotaWidget';

export default {
  name: 'MapLoadsQuota',
  components: {
    StackedQuotaWidget,
    QuotaContainer
  },
  data () {
    return {
    };
  },
  computed: {
    ...mapState({
      mapViewsQuota: state => state.user.map_views_quota,
      mapViews: state => state.user.map_views
    }),
    helpLink () {
      return 'https://carto.com/help/glossary/#mapload';
    }
  },
  methods: {
  }
};
</script>
