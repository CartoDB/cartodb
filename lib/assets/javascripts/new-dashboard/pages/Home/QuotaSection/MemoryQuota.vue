<template>
  <QuotaContainer :title="$t(`QuotaSection.memory`)" :perMonth="false">
    <StackedQuotaWidget
      :name="$t(`QuotaSection.yourPlan`)"
      :usedQuota="usedStorage"
      :availableQuota="getAmountInUnit(availableQuota)"
      :unit="getUnit(availableQuota)"
      :formatToLocale="false"
      helpText="TODO"
      :helpLink="storageHelpLink"/>
    <StackedQuotaWidget
      v-if="subscriptionsPremiumSizeInBytes"
      :name="$t(`QuotaSection.addOns`)"
      :usedQuota="usedPremiumDataSubscriptions"
      :unit="getUnit(subscriptionsPremiumSizeInBytes)"
      :formatToLocale="false"
      :helpLink="storageHelpLink" unlimited/>
  </QuotaContainer>
</template>

<script>
import { mapState } from 'vuex';
import QuotaContainer from './QuotaContainer';
import StackedQuotaWidget from './StackedQuotaWidget';
import { getAmountInUnit, getUnit, getExpBaseTwo } from 'new-dashboard/utils/storage-utils';

const DATA_SUBSCRIPTIONS_COLOR = '#2e51e8';
const DATASETS_COLOR = '#11a2b8';

export default {
  name: 'MemoryQuota',
  components: {
    StackedQuotaWidget,
    QuotaContainer
  },
  data () {
    return {
    };
  },
  computed: {
    ...mapState({
      quotaInBytes: state => state.user.storage.quota_in_bytes,
      dbSizeInBytes: state => state.user.storage.db_size_in_bytes,
      subscriptionsPublicSizeInBytes: state => state.user.storage.subscriptions_public_size_in_bytes,
      subscriptionsPremiumEstimatedSizeInBytes: state => state.user.storage.subscriptions_premium_estimated_size_in_bytes,
      subscriptionsPremiumSizeInBytes: state => state.user.storage.subscriptions_premium_size_in_bytes
    }),
    availableQuota () {
      return this.quotaInBytes - this.subscriptionsPremiumEstimatedSizeInBytes;
    },
    datasetsSize () {
      return Math.max(0, this.dbSizeInBytes - this.subscriptionsPremiumSizeInBytes - this.subscriptionsPublicSizeInBytes);
    },
    usedStorage () {
      return [{
        color: DATASETS_COLOR,
        label: 'Datasets',
        value: getAmountInUnit(this.datasetsSize, this.storageExponent)
      }, ...(this.subscriptionsPublicSizeInBytes ? [{
        color: DATA_SUBSCRIPTIONS_COLOR,
        label: 'Public data subscriptions',
        value: getAmountInUnit(this.subscriptionsPublicSizeInBytes, this.storageExponent)
      }] : [])];
    },
    usedPremiumDataSubscriptions () {
      return [
        {
          color: DATA_SUBSCRIPTIONS_COLOR,
          label: 'Premium data subscription',
          value: getAmountInUnit(this.subscriptionsPremiumSizeInBytes)
        }
      ];
    },
    storageExponent () {
      return getExpBaseTwo(this.quotaInBytes);
    },
    storageHelpLink () {
      return 'https://carto.com/help/your-account/your-disk-storage/';
    }
  },
  methods: {
    getAmountInUnit,
    getUnit
  }
};
</script>
