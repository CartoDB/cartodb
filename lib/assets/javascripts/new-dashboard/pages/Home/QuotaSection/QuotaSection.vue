<template>
  <div class="section is-bgSoftBlue section--noBorder">
    <div class="container">
      <div class="full-width">
        <SectionTitle class="grid-cell">
          <template slot="icon">
            <img svg-inline src="../../../assets/icons/section-title/quota.svg">
          </template>
        <template slot="title">{{ $t('QuotaSection.title') }}</template>
        </SectionTitle>
        <ul class="grid quota-list">
          <li class="grid-cell grid-cell--col12 quota-listitem">
            <AccountQuota></AccountQuota>
          </li>
          <li class="grid-cell grid-cell--col12 quota-listitem">
            <DataServicesQuota></DataServicesQuota>
          </li>
        </ul>
        <div class="quota-billing">
          <span class="quota-billingday text is-small is-txtSoftGrey">
            {{ $t(`QuotaSection.credits`, { day: billingDay })}}
          </span>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import SectionTitle from 'new-dashboard/components/SectionTitle';
import AccountQuota from './AccountQuota';
import DataServicesQuota from './DataServicesQuota';
import format from 'date-fns/format';

export default {
  name: 'QuotasModule',
  components: {
    SectionTitle,
    AccountQuota,
    DataServicesQuota
  },
  computed: {
    ...mapState({
      billingPeriod: state => state.user.next_billing_period
    }),
    billingDay () {
      return format(new Date(this.billingPeriod), 'Do');
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.quota-listitem {
  &:not(:first-of-type) {
    margin-top: 64px;
  }
}

.quota-billing {
  display: flex;
  justify-content: flex-end;
  padding: 0 10px;
}

.quota-billingday {
  margin-top: 16px;

  &::before {
    content: '*';
  }
}
</style>
