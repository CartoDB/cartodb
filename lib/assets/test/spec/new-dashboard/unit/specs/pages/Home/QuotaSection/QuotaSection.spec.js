import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import usersArray from '../../../fixtures/users';
import QuotaSection from 'new-dashboard/pages/Home/QuotaSection/QuotaSection';

const localVue = createLocalVue();
localVue.use(Vuex);

let store = new Vuex.Store({
  state: {
    user: usersArray[2]
  }
});

const $tc = key => key;
const $t = key => key;

describe('QuotaSection.vue', () => {
  it('should render correct contents', () => {
    const quotaSection = shallowMount(QuotaSection, {
      store,
      localVue,
      mocks: {
        $tc, $t
      }
    });

    expect(quotaSection).toMatchSnapshot();
  });
});
