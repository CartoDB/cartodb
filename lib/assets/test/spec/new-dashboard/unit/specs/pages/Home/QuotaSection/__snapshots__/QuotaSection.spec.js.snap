// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaSection.vue should render correct contents 1`] = `
<div class="section is-bgSoftBlue section--noBorder">
  <div class="container">
    <div class="full-width">
      <sectiontitle-stub class="grid-cell"><template></template> <template></template></sectiontitle-stub>
      <ul class="grid quota-list">
        <li class="grid-cell grid-cell--col12 quota-listitem">
          <accountquota-stub></accountquota-stub>
        </li>
        <li class="grid-cell grid-cell--col12 quota-listitem">
          <dataservicesquota-stub></dataservicesquota-stub>
        </li>
      </ul>
      <div class="quota-billing"><span class="quota-billingday text is-small is-txtSoftGrey">
          QuotaSection.credits
        </span></div>
    </div>
  </div>
</div>
`;
